src/core/execute.c

```diff
@@ -1728,8 +1729,46 @@ static int build_environment(
                 our_env[n_env++] = x;
         }
 
+        for (t = 0; t < _EXEC_DIRECTORY_TYPE_MAX; t++) {
+                _cleanup_free_ char *pre = NULL, *joined = NULL, *name = NULL;
+                const char *n;
+                char *d;
+
+                if (!p->prefix[t])
+                        continue;
+
+                if (strv_isempty(c->directories[t].paths))
+                        continue;
+
+                n = exec_directory_type_to_string(t);
```

Comment:
instead of deriving the env var name from exec_directory_type_to_string() (and even uppercasing it) I'd probably just have added an explicit, internal array here with the names, i.e.:

 ```c
static const char* env_var_names[_EXEC_DIRECTORY_TYPE_MAX] = {
        [EXEC_DIRECTORY_RUNTIME] = "RUNTIME_DIRECTORY", 
        …
};
```

After all, should we ever add another dir of this type, with a name of "DirectoryFooBar", then the env var name would be "DIRECTORY_FOO_BAR" and not "DIRECTORY_FOOBAR", i.e. a simple uppercasing wouldn't work in that case anyway...